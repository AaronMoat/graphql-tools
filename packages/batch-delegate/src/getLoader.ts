import { getNamedType, GraphQLOutputType, GraphQLList, GraphQLSchema, FieldNode } from 'graphql';

import DataLoader from 'dataloader';

import {
  SubschemaConfig,
  Transformer,
  createRequestFromInfo,
  getDelegationContext,
  getDelegatingOperation,
  getExecutor,
  validateRequest,
  Receiver,
  externalValueFromResult,
} from '@graphql-tools/delegate';
import { isAsyncIterable, relocatedError } from '@graphql-tools/utils';

import { BatchDelegateOptions } from './types';

const cache1: WeakMap<
  ReadonlyArray<FieldNode>,
  WeakMap<GraphQLSchema | SubschemaConfig, Record<string, DataLoader<any, any>>>
> = new WeakMap();

function createBatchFn<K = any>(options: BatchDelegateOptions) {
  const argsFromKeys = options.argsFromKeys ?? ((keys: ReadonlyArray<K>) => ({ ids: keys }));
  const { lazyOptionsFn } = options;

  return async (keys: ReadonlyArray<K>) => {
    const {
      context,
      info,
      operationName,
      operation = getDelegatingOperation(info.parentType, info.schema),
      fieldName = info.fieldName,
      returnType = new GraphQLList(getNamedType(options.info.returnType) as GraphQLOutputType),
      selectionSet,
      fieldNodes,
      binding,
      skipValidation,
    } = options;

    if (operation !== 'query' && operation !== 'mutation') {
      throw new Error(`Batch delegation not possible for operation '${operation}'.`);
    }

    const request = createRequestFromInfo({
      info,
      operation,
      fieldName,
      selectionSet,
      fieldNodes,
      operationName,
    });

    const delegationContext = getDelegationContext({
      request,
      args: argsFromKeys(keys),
      onLocatedError: originalError => relocatedError(originalError, originalError.path.slice(1)),
      ...(lazyOptionsFn == null ? options : lazyOptionsFn(options)),
      operation,
      fieldName,
      returnType,
    });

    const transformer = new Transformer(delegationContext, binding);

    const processedRequest = transformer.transformRequest(request);

    if (!skipValidation) {
      validateRequest(delegationContext, processedRequest.document);
    }

    const executor = getExecutor(delegationContext);

    const batchResult = await executor({
      ...processedRequest,
      context,
      info,
    });

    if (isAsyncIterable(batchResult)) {
      // TODO: split the asyncIterable and make a new receiver from each of them, return the Receiver instead of the
      // initial value, so that the correct info can be used to instantiate the Receiver
      const receiver = new Receiver(batchResult, delegationContext, executionResult =>
        transformer.transformResult(executionResult)
      );

      const batchValue = await receiver.getInitialValue();

      return Array.isArray(batchValue) ? batchValue : keys.map(() => batchValue);
    }

    // TODO: split the batchedResult and return the result instead of the value, so the correct info
    // can be used to instantiate the value
    const batchValue = externalValueFromResult(transformer.transformResult(batchResult), delegationContext);

    return Array.isArray(batchValue) ? batchValue : keys.map(() => batchValue);
  };
}

export function getLoader<K = any, V = any, C = K>(options: BatchDelegateOptions): DataLoader<K, V, C> {
  const fieldName = options.fieldName ?? options.info.fieldName;

  let cache2: WeakMap<GraphQLSchema | SubschemaConfig, Record<string, DataLoader<K, V, C>>> = cache1.get(
    options.info.fieldNodes
  );

  if (cache2 === undefined) {
    cache2 = new WeakMap();
    cache1.set(options.info.fieldNodes, cache2);
    const loaders = Object.create(null);
    cache2.set(options.schema, loaders);
    const batchFn = createBatchFn(options);
    const loader = new DataLoader<K, V, C>(keys => batchFn(keys), options.dataLoaderOptions);
    loaders[fieldName] = loader;
    return loader;
  }

  let loaders = cache2.get(options.schema);

  if (loaders === undefined) {
    loaders = Object.create(null);
    cache2.set(options.schema, loaders);
    const batchFn = createBatchFn(options);
    const loader = new DataLoader<K, V, C>(keys => batchFn(keys), options.dataLoaderOptions);
    loaders[fieldName] = loader;
    return loader;
  }

  let loader = loaders[fieldName];

  if (loader === undefined) {
    const batchFn = createBatchFn(options);
    loader = new DataLoader<K, V, C>(keys => batchFn(keys), options.dataLoaderOptions);
    loaders[fieldName] = loader;
  }

  return loader;
}
