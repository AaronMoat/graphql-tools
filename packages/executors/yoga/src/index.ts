import { GraphQLSchema } from 'graphql';
import { Plugin } from 'graphql-yoga';
import { useExecutor as useEnvelopExecutor } from '@graphql-tools/executor-envelop';
import { Executor, isPromise, MaybePromise } from '@graphql-tools/utils';
import { schemaFromExecutor } from '@graphql-tools/wrap';

type Opts = Parameters<typeof schemaFromExecutor>[2] & {
  polling?: number;
};

export function useExecutor(
  executor: Executor,
  opts?: Opts,
): Plugin & { invalidateSupergraph: () => void } {
  let schema$: MaybePromise<GraphQLSchema> | undefined;
  let schema: GraphQLSchema | undefined;
  if (opts?.polling) {
    setInterval(() => {
      schema$ = undefined;
      schema = undefined;
    }, opts.polling);
  }
  return {
    onPluginInit({ addPlugin }) {
      addPlugin(
        // @ts-expect-error TODO: fix typings
        useEnvelopExecutor(executor),
      );
    },
    onRequestParse({ serverContext }) {
      return {
        onRequestParseDone() {
          if (!schema$) {
            schema$ ||= schemaFromExecutor(executor, serverContext, opts);
            if (isPromise(schema$)) {
              return schema$.then(newSchema => {
                schema = newSchema;
              }) as Promise<void>;
            }
          }
        },
      };
    },
    onEnveloped({ setSchema }) {
      if (!schema) {
        throw new Error(
          `You provide a promise of a schema but it hasn't been resolved yet. Make sure you use this plugin with GraphQL Yoga.`,
        );
      }
      setSchema(schema);
    },
    invalidateSupergraph() {
      schema$ = undefined;
      schema = undefined;
    },
  };
}
