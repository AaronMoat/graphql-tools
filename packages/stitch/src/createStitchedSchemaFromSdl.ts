/* eslint-disable no-eval */
import { buildASTSchema, DocumentNode, parse, SchemaDefinitionNode, visit } from 'graphql';
import { MergedTypeConfig, SubschemaConfig } from '@graphql-tools/delegate';
import { Executor } from '@graphql-tools/utils';
import { stitchSchemas } from './stitchSchemas.js';

export function createStichedSchemaFromSdl(
  sdl: string | DocumentNode,
  subschemaExecutorMap: Map<string, Executor>,
) {
  const documentNode = typeof sdl === 'string' ? parse(sdl) : sdl;
  const { subschemaNames } = extractSubschemaNamesAndExecutors(documentNode);
  const subschemas: SubschemaConfig[] = [];
  for (const subschemaName of subschemaNames) {
    let mergedTypeConfigMap: Record<string, MergedTypeConfig> | undefined;
    const schemaAst = visit(documentNode, {
      ObjectTypeDefinition: {
        enter(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'subschemaObjectType') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                const implementsArg = directiveNode.arguments?.find(
                  arg => arg.name.value === 'implements',
                );
                const implementsArgValue = implementsArg?.value;
                if (implementsArgValue?.kind === 'ListValue') {
                  const interfaces = implementsArgValue.values.map(value => {
                    if (value.kind === 'StringValue') {
                      return value.value;
                    }
                    throw new Error(`Expected interface name to be a string, found ${value.kind}`);
                  });
                  return {
                    ...node,
                    interfaces: node.interfaces?.filter(iface =>
                      interfaces.includes(iface.name.value),
                    ),
                  };
                }
                return node;
              }
            }
          }
          return null;
        },
        leave(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'merge') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                const mergedTypeConfig: MergedTypeConfig = {};
                for (const argumentNode of directiveNode.arguments ?? []) {
                  if (argumentNode.name.value === 'selectionSet') {
                    const selectionSetArgValue = argumentNode.value;
                    if (selectionSetArgValue.kind === 'StringValue') {
                      mergedTypeConfig.selectionSet = selectionSetArgValue.value;
                    }
                  } else if (argumentNode.name.value === 'fieldName') {
                    const fieldNameArgValue = argumentNode.value;
                    if (fieldNameArgValue.kind === 'StringValue') {
                      mergedTypeConfig.fieldName = fieldNameArgValue.value;
                    }
                  } else if (argumentNode.name.value === 'key') {
                    const keyArgValue = argumentNode.value;
                    if (keyArgValue.kind === 'StringValue') {
                      mergedTypeConfig.key = eval(keyArgValue.value) as any;
                    }
                  } else if (argumentNode.name.value === 'argsFromKeys') {
                    const argsFromKeysArgValue = argumentNode.value;
                    if (argsFromKeysArgValue.kind === 'StringValue') {
                      mergedTypeConfig.argsFromKeys = eval(argsFromKeysArgValue.value) as any;
                    }
                  }
                }
                const typeName = node.name.value;
                mergedTypeConfigMap ||= {};
                mergedTypeConfigMap[typeName] = mergedTypeConfig;
              }
            }
          }
        },
      },
      InterfaceTypeDefinition: {
        enter(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'subschemaInterfaceType') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                const implementsArg = directiveNode.arguments?.find(
                  arg => arg.name.value === 'implements',
                );
                const implementsArgValue = implementsArg?.value;
                if (implementsArgValue?.kind === 'ListValue') {
                  const interfaces = implementsArgValue.values.map(value => {
                    if (value.kind === 'StringValue') {
                      return value.value;
                    }
                    throw new Error(`Expected interface name to be a string, found ${value.kind}`);
                  });
                  return {
                    ...node,
                    interfaces: node.interfaces?.filter(
                      iface => !interfaces.includes(iface.name.value),
                    ),
                  };
                }
                return node;
              }
            }
          }
          return null;
        },
      },
      InputObjectTypeDefinition: {
        enter(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'subschemaInputObjectType') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                return node;
              }
            }
          }
          return null;
        },
      },
      FieldDefinition: {
        enter(node) {
          for (const directiveNode of node.directives || []) {
            if (
              directiveNode.name.value === 'subschemaObjectField' ||
              directiveNode.name.value === 'subschemaInterfaceField' ||
              directiveNode.name.value === 'subschemaInputObjectField'
            ) {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                return {
                  ...node,
                  arguments: node.arguments?.filter(arg => {
                    const argSubschemaDirective = arg.directives?.find(
                      directive => directive.name.value === 'subschemaArgument',
                    );
                    const argSubschemaArg = argSubschemaDirective?.arguments?.find(
                      arg => arg.name.value === 'subschema',
                    );
                    const argSubschemaArgValue = argSubschemaArg?.value;
                    if (
                      argSubschemaArgValue?.kind === 'StringValue' &&
                      argSubschemaArgValue.value === subschemaName
                    ) {
                      return true;
                    }
                    return false;
                  }),
                };
              }
            }
          }
          return null;
        },
      },
      UnionTypeDefinition: {
        enter(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'subschemaUnionType') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                const typesArg = directiveNode.arguments?.find(arg => arg.name.value === 'types');
                const typesArgValue = typesArg?.value;
                if (typesArgValue?.kind === 'ListValue') {
                  const types = typesArgValue.values.map(value => {
                    if (value.kind === 'StringValue') {
                      return value.value;
                    }
                    throw new Error(`Expected type name to be a string, found ${value.kind}`);
                  });
                  return {
                    ...node,
                    types: node.types?.filter(type => !types.includes(type.name.value)),
                  };
                }
              }
            }
          }
          return null;
        },
      },
      ScalarTypeDefinition: {
        enter(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'subschemaScalarType') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                return node;
              }
            }
          }
          return null;
        },
      },
      EnumTypeDefinition: {
        enter(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'subschemaEnumType') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                return node;
              }
            }
          }
          return null;
        },
      },
      EnumValueDefinition: {
        enter(node) {
          for (const directiveNode of node.directives ?? []) {
            if (directiveNode.name.value === 'subschemaEnumValue') {
              const subschemaArg = directiveNode.arguments?.find(
                arg => arg.name.value === 'subschema',
              );
              const subschemaArgValue = subschemaArg?.value;
              if (
                subschemaArgValue?.kind === 'StringValue' &&
                subschemaArgValue.value === subschemaName
              ) {
                return node;
              }
            }
          }
          return null;
        },
      },
    });
    const schema = buildASTSchema(schemaAst, {
      assumeValid: true,
      assumeValidSDL: true,
    });
    subschemas.push({
      name: subschemaName,
      schema,
      merge: mergedTypeConfigMap,
      executor: subschemaExecutorMap.get(subschemaName),
    });
  }
  return stitchSchemas({
    subschemas,
  });
}

function extractSubschemaNamesAndExecutors(documentNode: DocumentNode) {
  const schemaDef = documentNode.definitions.find(
    def => def.kind === 'SchemaDefinition',
  ) as SchemaDefinitionNode;
  const subschemaNames = new Set<string>();
  const subschemaNameExecutorMap = new Map<string, string>();
  if (schemaDef) {
    schemaDef.directives?.forEach(directive => {
      if (directive.name.value === 'subschema') {
        const subschemaNameArg = directive.arguments?.find(arg => arg.name.value === 'subschema');
        if (subschemaNameArg) {
          const subschemaNameArgValue = subschemaNameArg.value;
          if (subschemaNameArgValue.kind === 'StringValue') {
            const subschemaName = subschemaNameArgValue.value;
            subschemaNames.add(subschemaName);
            const executorArg = directive.arguments?.find(arg => arg.name.value === 'executor');
            if (executorArg) {
              const executorArgValue = executorArg.value;
              if (executorArgValue.kind === 'StringValue') {
                subschemaNameExecutorMap.set(subschemaName, executorArgValue.value);
              } else {
                throw new Error(`Expected executor to be a string, found ${executorArgValue.kind}`);
              }
            }
          } else {
            throw new Error(
              `Expected subschema name to be a string, found ${subschemaNameArgValue.kind}`,
            );
          }
        }
      }
    });
  }
  return { subschemaNames, subschemaNameExecutorMap };
}
