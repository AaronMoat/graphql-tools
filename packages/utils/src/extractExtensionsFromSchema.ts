import { GraphQLFieldConfig, GraphQLSchema } from 'graphql';
import { asArray } from './helpers.js';
import { MapperKind } from './Interfaces.js';
import { mapSchema } from './mapSchema.js';
import {
  EnumTypeExtensions,
  InputTypeExtensions,
  ObjectTypeExtensions,
  SchemaExtensions,
} from './types.js';

function handleDirectiveExtensions(extensions: any, removeDirectives: boolean) {
  extensions = extensions || {};
  const { directives: existingDirectives, ...rest } = extensions;
  const finalExtensions: any = {
    ...rest,
  };
  if (!removeDirectives) {
    if (existingDirectives != null) {
      const directives = {};
      for (const directiveName in existingDirectives) {
        directives[directiveName] = [...asArray(existingDirectives[directiveName])];
      }
      finalExtensions.directives = directives;
    }
  }
  return finalExtensions;
}

export function extractExtensionsFromSchema(
  schema: GraphQLSchema,
  removeDirectives = false,
): SchemaExtensions {
  const result: SchemaExtensions = {
    schemaExtensions: handleDirectiveExtensions(schema.extensions, removeDirectives),
    types: {},
  };

  mapSchema(schema, {
    [MapperKind.OBJECT_TYPE]: type => {
      result.types[type.name] = {
        fields: {},
        type: 'object',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.INTERFACE_TYPE]: type => {
      result.types[type.name] = {
        fields: {},
        type: 'interface',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.FIELD]: (field, fieldName, typeName) => {
      (result.types[typeName] as ObjectTypeExtensions).fields[fieldName] = {
        arguments: {},
        extensions: handleDirectiveExtensions(field.extensions, removeDirectives),
      };
      const args = (field as GraphQLFieldConfig<any, any>).args;
      if (args != null) {
        for (const argName in args) {
          (result.types[typeName] as ObjectTypeExtensions).fields[fieldName].arguments[argName] =
            handleDirectiveExtensions(args[argName].extensions, removeDirectives);
        }
      }
      return field;
    },
    [MapperKind.ENUM_TYPE]: type => {
      result.types[type.name] = {
        values: {},
        type: 'enum',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.ENUM_VALUE]: (value, typeName, _schema, valueName) => {
      (result.types[typeName] as EnumTypeExtensions).values[valueName] = handleDirectiveExtensions(
        value.extensions,
        removeDirectives,
      );
      return value;
    },
    [MapperKind.SCALAR_TYPE]: type => {
      result.types[type.name] = {
        type: 'scalar',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.UNION_TYPE]: type => {
      result.types[type.name] = {
        type: 'union',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.INPUT_OBJECT_TYPE]: type => {
      result.types[type.name] = {
        fields: {},
        type: 'input',
        extensions: handleDirectiveExtensions(type.extensions, removeDirectives),
      };
      return type;
    },
    [MapperKind.INPUT_OBJECT_FIELD]: (field, fieldName, typeName) => {
      (result.types[typeName] as InputTypeExtensions).fields[fieldName] = {
        extensions: handleDirectiveExtensions(field.extensions, removeDirectives),
      };
      return field;
    },
  });

  return result;
}
