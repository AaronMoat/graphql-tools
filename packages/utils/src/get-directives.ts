import {
  GraphQLEnumTypeConfig,
  GraphQLEnumValue,
  GraphQLEnumValueConfig,
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLInputField,
  GraphQLInputFieldConfig,
  GraphQLInputObjectTypeConfig,
  GraphQLInterfaceTypeConfig,
  GraphQLNamedType,
  GraphQLObjectTypeConfig,
  GraphQLScalarTypeConfig,
  GraphQLSchema,
  GraphQLSchemaConfig,
  GraphQLUnionTypeConfig,
} from 'graphql';
import { getDirectiveExtensions } from './getDirectiveExtensions.js';

export interface DirectiveAnnotation {
  name: string;
  args?: Record<string, any>;
}

export type DirectableGraphQLObject =
  | GraphQLSchema
  | GraphQLSchemaConfig
  | GraphQLNamedType
  | GraphQLObjectTypeConfig<any, any>
  | GraphQLInterfaceTypeConfig<any, any>
  | GraphQLUnionTypeConfig<any, any>
  | GraphQLScalarTypeConfig<any, any>
  | GraphQLEnumTypeConfig
  | GraphQLEnumValue
  | GraphQLEnumValueConfig
  | GraphQLInputObjectTypeConfig
  | GraphQLField<any, any>
  | GraphQLInputField
  | GraphQLFieldConfig<any, any>
  | GraphQLInputFieldConfig;

export function getDirectivesInExtensions(
  node: DirectableGraphQLObject,
  pathToDirectivesInExtensions = ['directives'],
): Array<DirectiveAnnotation> {
  const directiveExtensions = getDirectiveExtensions(node, undefined, pathToDirectivesInExtensions);
  return Object.entries(directiveExtensions)
    .map(([directiveName, directiveArgsArr]) =>
      directiveArgsArr?.map(directiveArgs => ({
        name: directiveName,
        args: directiveArgs,
      })),
    )
    .flat(Infinity)
    .filter(Boolean) as Array<DirectiveAnnotation>;
}

export function getDirectiveInExtensions(
  node: DirectableGraphQLObject,
  directiveName: string,
  pathToDirectivesInExtensions = ['directives'],
): Array<Record<string, any>> | undefined {
  const directiveExtensions = getDirectiveExtensions(node, undefined, pathToDirectivesInExtensions);
  return directiveExtensions[directiveName];
}

export function getDirectives(
  schema: GraphQLSchema,
  node: DirectableGraphQLObject,
  pathToDirectivesInExtensions = ['directives'],
): Array<DirectiveAnnotation> {
  const directiveExtensions = getDirectiveExtensions(node, schema, pathToDirectivesInExtensions);
  return Object.entries(directiveExtensions)
    .map(([directiveName, directiveArgsArr]) =>
      directiveArgsArr?.map(directiveArgs => ({
        name: directiveName,
        args: directiveArgs,
      })),
    )
    .flat(Infinity)
    .filter(Boolean) as Array<DirectiveAnnotation>;
}

export function getDirective(
  schema: GraphQLSchema,
  node: DirectableGraphQLObject,
  directiveName: string,
  pathToDirectivesInExtensions = ['directives'],
): Array<Record<string, any>> | undefined {
  const directiveExtensions = getDirectiveExtensions(node, schema, pathToDirectivesInExtensions);
  return directiveExtensions[directiveName];
}
