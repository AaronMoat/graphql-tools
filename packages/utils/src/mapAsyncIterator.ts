import type { MaybePromise } from './executor.js';
import { isAsyncIterable } from './isAsyncIterable.js';
import { isPromise } from './jsutils.js';

/**
 * Given an AsyncIterable and a callback function, return an AsyncIterator
 * which produces values mapped via calling the callback function.
 */
export function mapAsyncIterator<T, U>(
  iteratorOrIterable: AsyncIterable<T> | AsyncIterator<T>,
  onNext: (value: T) => MaybePromise<U>,
  onError?: (error: unknown) => unknown,
  onEnd?: () => MaybePromise<void>,
): AsyncIterableIterator<U> {
  let _mappedIterator: AsyncIterator<U>;

  function getMappedIterator() {
    if (_mappedIterator == null) {
      let _iterator: AsyncIterator<T>;
      let abruptClose: (error: unknown) => Promise<never>;
      function getIterator() {
        if (!_iterator) {
          _iterator = isAsyncIterable(iteratorOrIterable)
            ? iteratorOrIterable[Symbol.asyncIterator]()
            : iteratorOrIterable;
          if (typeof _iterator.return === 'function') {
            const $return = _iterator.return;
            abruptClose = (error: any) => {
              const rethrow = () => Promise.reject(error);
              return $return.call(_iterator).then(rethrow, rethrow);
            };
          }
        }
        return _iterator;
      }
      let onEndWithValue: <R>(value: R) => MaybePromise<R>;

      if (onEnd) {
        onEndWithValue = value => {
          const onEnd$ = onEnd();
          return isPromise(onEnd$) ? onEnd$.then(() => value) : value;
        };
      }

      function mapResult(result: any) {
        if (result.done) {
          return onEndWithValue ? onEndWithValue(result) : result;
        }
        return asyncMapValue(result.value, onNext).then(iteratorResult, abruptClose);
      }

      let mapReject: (error: unknown) => Promise<unknown>;
      if (onError) {
        // Capture rejectCallback to ensure it cannot be null.
        const reject = onError;
        mapReject = error => asyncMapValue(error, reject).then(iteratorResult, abruptClose);
      }
      _mappedIterator = {
        next(value) {
          return getIterator().next(value).then(mapResult, mapReject);
        },
        return(value) {
          const iterator = getIterator();
          const res$ = iterator.return
            ? iterator.return(value).then(mapResult, mapReject)
            : Promise.resolve({ value: undefined, done: true });
          return onEndWithValue ? res$.then(onEndWithValue) : res$;
        },
        throw(error) {
          const iterator = getIterator();
          if (typeof iterator.throw === 'function') {
            return iterator.throw(error).then(mapResult, mapReject);
          }
          return Promise.reject(error).catch(abruptClose);
        },
      };
    }
    return _mappedIterator;
  }

  return {
    get next() {
      return getMappedIterator().next;
    },
    get return() {
      return getMappedIterator().return;
    },
    get throw() {
      return getMappedIterator().throw;
    },
    [Symbol.asyncIterator]() {
      return this;
    },
  };
}

function asyncMapValue<T, U>(value: T, callback: (value: T) => PromiseLike<U> | U): Promise<U> {
  return new Promise(resolve => resolve(callback(value)));
}

function iteratorResult<T>(value: T): IteratorResult<T> {
  return { value, done: false };
}
