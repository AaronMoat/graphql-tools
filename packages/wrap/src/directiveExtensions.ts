export interface DirectiveExtensionsTarget {
  extensions?: {
    directives?: Record<string, any>;
  } | null;
}

export function addDirectiveExtensions(
  target: DirectiveExtensionsTarget,
  newDirectives: Record<string, any>,
) {
  target.extensions ||= {};
  target.extensions.directives ||= {};
  const targetDirectives = target.extensions.directives;
  for (const directiveName in newDirectives) {
    const existingDirective = targetDirectives[directiveName];
    const newDirective = newDirectives[directiveName];
    if (!existingDirective) {
      targetDirectives[directiveName] = newDirective;
    } else if (Array.isArray(existingDirective)) {
      if (Array.isArray(newDirective)) {
        targetDirectives[directiveName] = existingDirective.concat(newDirective);
      } else {
        targetDirectives[directiveName] = [...existingDirective, newDirective];
      }
    } else {
      if (Array.isArray(newDirective)) {
        targetDirectives[directiveName] = [existingDirective, ...newDirective];
      } else {
        targetDirectives[directiveName] = [existingDirective, newDirective];
      }
    }
  }
}
