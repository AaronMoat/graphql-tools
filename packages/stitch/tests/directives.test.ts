import { makeExecutableSchema } from '@graphql-tools/schema';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { stitchSchemas } from '../src/stitchSchemas';

describe('Directives', () => {
  it('adds resolveTo directive to delegated root fields', async () => {
    const users = [
      {
        id: '0',
        name: 'Ada Lovelace',
      },
      {
        id: '1',
        name: 'Alan Turing',
      },
    ];
    const stitchedSchema = stitchSchemas({
      subschemas: [
        {
          name: 'Users',
          schema: makeExecutableSchema({
            typeDefs: /* GraphQL */ `
              type Query {
                users(ids: [ID!]!): [User!]!
                user(id: ID!): User!
              }

              type Mutation {
                addUser(user: UserInput!): User!
              }

              input UserInput {
                name: String!
              }

              interface Node {
                id: ID!
              }

              type User implements Node {
                id: ID!
                name: String!
              }
            `,
            resolvers: {
              Query: {
                users: (_, { ids }) => users.filter(user => ids.includes(user.id)),
                user: (_, { id }) => users.find(user => user.id === id),
              },
            },
          }),
        },
        {
          name: 'Posts',
          schema: makeExecutableSchema({
            typeDefs: /* GraphQL */ `
              type Query {
                posts(ids: [ID!]!): [Post!]!
                post(id: ID!): Post!
              }

              type Mutation {
                addPost(post: PostInput!): Post!
              }

              input PostInput {
                title: String!
                authorId: ID!
              }

              interface Node {
                id: ID!
              }

              type Post implements Node {
                id: ID!
                title: String!
                authorId: ID!
              }
            `,
            resolvers: {
              Query: {
                posts: () => [
                  {
                    id: '0',
                    title: 'Hello World',
                    authorId: '0',
                  },
                  {
                    id: '1',
                    title: 'Foo Bar',
                    authorId: '1',
                  },
                ],
              },
            },
          }),
        },
      ],
    });
    expect(printSchemaWithDirectives(stitchedSchema)).toMatchSnapshot('schema-with-directives');
  });
});
