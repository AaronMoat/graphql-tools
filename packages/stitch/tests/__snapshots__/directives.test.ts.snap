// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Directives adds resolveTo directive to delegated root fields: schema-with-directives 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

type Query @subschemaObjectType(subschema: "Users", type: "Query") @subschemaObjectType(subschema: "Posts", type: "Query") {
  users(ids: [ID!]! @subschemaArgument(subschema: "Users", type: "Query", field: "users")): [User!]! @resolveTo(subschema: "Users", operation: "query") @subschemaRootField(subschema: "Users", type: "Query", field: "users")
  user(id: ID! @subschemaArgument(subschema: "Users", type: "Query", field: "user")): User! @resolveTo(subschema: "Users", operation: "query") @subschemaRootField(subschema: "Users", type: "Query", field: "user")
  posts(ids: [ID!]! @subschemaArgument(subschema: "Posts", type: "Query", field: "posts")): [Post!]! @resolveTo(subschema: "Posts", operation: "query") @subschemaRootField(subschema: "Posts", type: "Query", field: "posts")
  post(id: ID! @subschemaArgument(subschema: "Posts", type: "Query", field: "post")): Post! @resolveTo(subschema: "Posts", operation: "query") @subschemaRootField(subschema: "Posts", type: "Query", field: "post")
}

type Mutation @subschemaObjectType(subschema: "Users", type: "Mutation") @subschemaObjectType(subschema: "Posts", type: "Mutation") {
  addUser(user: UserInput! @subschemaArgument(subschema: "Users", type: "Mutation", field: "addUser")): User! @resolveTo(subschema: "Users", operation: "mutation") @subschemaRootField(subschema: "Users", type: "Mutation", field: "addUser")
  addPost(post: PostInput! @subschemaArgument(subschema: "Posts", type: "Mutation", field: "addPost")): Post! @resolveTo(subschema: "Posts", operation: "mutation") @subschemaRootField(subschema: "Posts", type: "Mutation", field: "addPost")
}

input UserInput @subschemaInputObjectType(subschema: "Users", type: "UserInput") {
  name: String! @subschemaInputObjectField(subschema: "Users", type: "UserInput", field: "name")
}

interface Node @subschemaInterfaceType(subschema: "Users", type: "Node") @subschemaInterfaceType(subschema: "Posts", type: "Node") {
  id: ID! @subschemaInterfaceField(subschema: "Users", type: "Node", field: "id") @subschemaInterfaceField(subschema: "Posts", type: "Node", field: "id")
}

type User implements Node @subschemaObjectType(subschema: "Users", type: "User", implements: ["Node"]) {
  id: ID! @subschemaObjectField(subschema: "Users", type: "User", field: "id")
  name: String! @subschemaObjectField(subschema: "Users", type: "User", field: "name")
}

input PostInput @subschemaInputObjectType(subschema: "Posts", type: "PostInput") {
  title: String! @subschemaInputObjectField(subschema: "Posts", type: "PostInput", field: "title")
  authorId: ID! @subschemaInputObjectField(subschema: "Posts", type: "PostInput", field: "authorId")
}

type Post implements Node @subschemaObjectType(subschema: "Posts", type: "Post", implements: ["Node"]) {
  id: ID! @subschemaObjectField(subschema: "Posts", type: "Post", field: "id")
  title: String! @subschemaObjectField(subschema: "Posts", type: "Post", field: "title")
  authorId: ID! @subschemaObjectField(subschema: "Posts", type: "Post", field: "authorId")
}"
`;
